### PR TITLE
feat[rust]: introduce `how="right"` parameter to join methods

### LINKED ISSUE #3934: Support right (outer) joins in pl.LazyFrame
#### Describe your feature request

Currently, `pl.LazyFrame` supports the following types of joins:

```
“inner” “left” “outer” “asof”, “cross” “semi” “anti”
```

One type of join that is missing from here is a right join. Of course, one could just join the two lazyframes in the other order:

```
lf1.join(lf2, how="right", ...) == lf2.join(lf1, how="left", ...)
```

but often times this expression is chained in a long lazy transformation which makes the above^ substitution harder to perform. I took a look at [the join implementation](https://github.com/pola-rs/polars/blob/master/py-polars/polars/internals/lazy_frame.py#L1248) and it seems relatively simple to swap `lf1, lf2` within based on `how`. Let me know if I'm missing anything, thanks!

### PR DESCRIPTION
**Pull request by matteosantama**
Closes #3934 

This MR has two commits

1. Simplify `DataFrame.join` and `LazyFrame.join` by housing all the "business logic" within `LazyFrame.join`. `DataFrame.join` then becomes a wrapper to `LazyFrame.join`. @ritchie46 any red flags here?
2. Adds `how="right"` parameter to both join methods.

Question: we call `collect(no_optimization=True)`, is there a reason for this? Or can we turn it on?

### CODE REVIEW

**Comment by ritchie46** [comment]
> Question: we call collect(no_optimization=True), is there a reason for this? Or can we turn it on?

Eager operations should not run an optimization pass, as they only contain a single node.

> Adds how="right" parameter to both join methods. 

Does this also keep the right hand columns on the right hand side?

File: py-polars/polars/internals/dataframe/frame.py [hunk 1]
```diff
@@ -3626,52 +3626,23 @@ def join(
...
-                    left_on,
-                    right_on,
-                    on=on,
-                    how=how,
-                    suffix=suffix,
-                )
-                .collect(no_optimization=True)
-            )
-        else:
-            return self._from_pydf(
```
Comment (ritchie46) [inline py-polars/polars/internals/dataframe/frame.py:3672]: I think we can then remove the `fn join` from the eager python bindings.

**Comment by matteosantama** [comment]
> Does this also keep the right hand columns on the right hand side?

Sort of... this is from the tests

```python
x = pl.DataFrame({"row_id": [1, 2, 3, 4, 5], "values": [5, 6, 7, 8, 9]})
y = pl.DataFrame({"row_id": [1, 2, 3, 4], "values": [10, 11, 12, 13]})

y.join(x, how="right", on="row_id", suffix="_x")
pl.DataFrame(
    {
        "values": [10, 11, 12, 13, None],
        "row_id": [1, 2, 3, 4, 5],
        "values_x": [5, 6, 7, 8, 9],
    }
)
```

The resulting dataframe has the `on` column in the middle. I will push a fix since this is not consistent with `how="left"`

**Comment by matteosantama** [comment]
@ritchie46 thinking about this more, I think the current behavior is in fact correct. The index column in preserved in the **right** `DataFrame`, which makes sense for a right merge. Consider the following:

```python
x = pl.DataFrame({"row_id_x": [1, 2, 3], "values": [4, 5, 6]})
y = pl.DataFrame({"row_id_y": [1, 2], "values": [7, 8]})

y.join(x, how="right", left_on="row_id_y", right_on="row_id_x", suffix="_x")
# shape: (3, 3)
# ┌────────┬──────────┬──────────┐
# │ values ┆ row_id_x ┆ values_x │
# │ ---    ┆ ---      ┆ ---      │
# │ i64    ┆ i64      ┆ i64      │
# ╞════════╪══════════╪══════════╡
# │ 7      ┆ 1        ┆ 4        │
# ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
# │ 8      ┆ 2        ┆ 5        │
# ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
# │ null   ┆ 3        ┆ 6        │
# └────────┴──────────┴──────────┘
```

Should we instead produce

```python
# shape: (3, 3)
# ┌──────────┬────────┬──────────┐
# │ row_id_x ┆ values ┆ values_x │
# │ ---      ┆ ---    ┆ ---      │
# │ i64      ┆ i64    ┆ i64      │
# ╞══════════╪════════╪══════════╡
# │ 1        ┆ 7      ┆ 4        │
# ├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
# │ 2        ┆ 8      ┆ 5        │
# ├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤
# │ 3        ┆ null   ┆ 6        │
# └──────────┴────────┴──────────┘
```

where the `_x` columns have been split up? I think not

**Comment by matteosantama** [comment]
Closing this -- I plan to split it up into separate, more digestible PRs

**Comment by ritchie46** [comment]
Note that I am really hesitant to introduce this feature as there the optimizers assume left side joins. 

This will likely introduce unforseen bugs that I want thoroughly tested before adding this. I think we should add this as an experimental feature and try to break it.

**Pull request closed without merging**